### ISSUE TITLE
[Debug] WinDbg and SCCPPass leads to variables being "optimized away" in certain cases.

### ISSUE THREAD

**Issue by tru** [llvm/llvm-project#55958]
I have been tracking an issue where a lot of our developers report that variables are usually not inspectable when in lambdas or other smaller scopes. We reduced one of the problem to the following test case:

```
#ifdef __clang__
#define OPTIMIZE_OFF _Pragma("clang optimize off") 
#define OPTIMIZE_ON  _Pragma("clang optimize on")
#else
#define OPTIMIZE_OFF __pragma(optimize("",off))
#define OPTIMIZE_ON  __pragma(optimize("", on))
#endif

OPTIMIZE_OFF
void foo(int param)
{
	(void)param;
}
OPTIMIZE_ON

int main()
{
	[&](int param)
	{
		param++;
		foo(param); // place breakpoint here and check 'param'
	}(1234);
}
```

If you break inside the lambda here you'll see that you can't inspect the variable `param`.

With MSVC I build with `/O2 /Ob0 /Zi` and I can break inside the lambda and inspect `param`. With clang-cl I build with the same flags and when breaking there it says the variable is "optimized away".

The PDB information is pretty similar for both clang and MSVC - here is the output of `llvm-pdbutil dump --symbols`:

Clang: [full output](https://gist.github.com/tru/9b618a8987c31b4c0b1fe5715df92fa0)
```
     352 | S_LOCAL [size = 16] `this`
           type=0x100A (const main::<lambda_0>*), flags = param | optimized away
     368 | S_LOCAL [size = 16] `param`
           type=0x0074 (int), flags = param | optimized away
```

Note how it's flagged to be `optimized away`. Both for `param` and `this`. I specifically investigated `param` here - but the fact that `this` is not available when breaking is also a bit annoying.

MSVC: [full output](https://gist.github.com/tru/3ccd3c8ae51a4c1a0f78e0d10482c826)
```
     508 | S_REGREL32 [size = 20] `this`
           type = 0x10F1 (const main::__l2::<lambda_865e41...), register = RSP, offset = 8
     528 | S_REGREL32 [size = 20] `param`
           type = 0x0074 (int), register = RSP, offset = 16
```

I also inspected with the new llvm-dva util:

Clang
```
Totals by lexical level:

             {CompileUnit} 'simple-lambda.cpp'
               {Function} extern not_inlined 'foo' -> 'void'
                 {Parameter} 'param' -> 'int'
    11           {Line}
                 {Code} 'pushq	%rax'
                 {Code} 'movl	%ecx, 0x4(%rsp)'
    13           {Line}
                 {Code} 'popq	%rax'
                 {Code} 'retq'
               {Function} extern not_inlined 'main' -> 'int'
    18           {InlinedFunction} inlined '::18' -> '<unnamed-tag> __scrt_narrow_environment_policy'
                   {Parameter} 'param' -> 'int'
    17           {Line}
                 {Code} 'subq	$0x28, %rsp'
    21           {Line}
                 {Code} 'movl	$0x4d3, %ecx'
                 {Code} 'callq	-0x1e'
    23           {Line}
                 {Code} 'xorl	%eax, %eax'
                 {Code} 'addq	$0x28, %rsp'
                 {Code} 'retq'
```

MSVC:
```
             {CompileUnit} 'simple-lambda.cpp'
               {Function} extern not_inlined 'foo' -> 'int'
                 {Variable} 'param' -> 'int'
    11           {Line}
                 {Code} 'movl	%ecx, 0x8(%rsp)'
    13           {Line}
                 {Code} 'retq'
               {Function} extern not_inlined 'main' -> 'int'
                 {InlinedFunction} inlined '::22' -> 'void'
    17           {Line}
                 {Code} 'subq	$0x28, %rsp'
    18           {Line}
                 {Code} 'movl	$0x4d3, %ecx'
                 {Code} 'callq	-0x1e'
    23           {Line}
                 {Code} 'xorl	%eax, %eax'
                 {Code} 'addq	$0x28, %rsp'
                 {Code} 'retq'
               {Function} inlined 'operator()' -> '__int64'
                 {Parameter} 'param' -> 'int'
                 {Variable} 'param' -> 'int'
    22           {Line}
    20           {Line}
                 {Code} 'leal	0x1(%rdx), %ecx'
    21           {Line}
                 {Code} 'jmp	0x8'
```

The one thing that sticks out here is the fact that in MSVC it shows up `param` as a variable - but I am not that adept at reading the output of `llvm-dva` yet.

Anyway that lead me to look at the actual IR/ASM to see if there was a huge difference. The thing that struck me was that LLVM converts the `param` bits to a constant. You can look at the compiler explorer output here: https://godbolt.org/z/7ffoYaEEc

With some `--print-after-all` the pass that converted the variable into a constant is the SCCPPass - so I disabled that pass just to see if it fixed the problem and it did. With that pass disabled I was able to debug the `param` variable correctly.

I have also tested opening the same file with lldb on Linux and in there I can debug the variable just fine.

This leads me to think that this is either
1) a problem with WinDbg and the fact that MSVC doesn't do constant folding here might just lead to WinDbg not really handling this case?
2) something in the PDB writer in LLVM. Maybe the "optimized away" flag is added to often?

Interested in hearing other people's input on this: @zmodem @joker-eph @djtodoro

**Comment by dwblaikie** [comment]
@rnk - looks like this might be some expressive limitation/unimplemented issue in CodeView, since this doesn't seem to be a problem for DWARF/lldb?

**Comment by rnk** [comment]
Yes, we noticed this, it's a super common issue with LLVM. Variables are very often initialized to constants (think zero init, `Foo *p = nullptr; int y = 0;`), which become dbg.values of constant expressions. This really goes all the way back to LLVM's design choice not to represent constants as IR instructions. Constant variables are often actually used by nearby instructions, so there is an instruction in the stream to materialize `1234`, but it is no longer associated with the variable.

Without changing that LLVM design choice or the CodeView format, the only option we have left is to heroically search the instruction stream for materializations of the constant that overlap with the range over which the variable is known to be that constant.

Basically, it's not feasible to fix this in the medium term. We could document the limitation or file a feature request with Microsoft, which is probably our best bet.

**Comment by rnk** [comment]
I don't think this issue is limited to SCCP either. SROA can do this pretty easily, set up examples like this:

```
int bar(int a, int b, int c);
void foo() {
  int a = 1;
  int b = 2;
  int c = 3;
  bar(a, b, c); // Surely abc will not be available in LLVM, I haven't checked MSVC's behavior, but it might work.
}

**Comment by tru** [comment]
If that's the case I wonder if it makes sense to have a flag to disable certain passes (SCCP and SROA? Can you think of any others) that can improve debugability in release builds? We can't really use debug builds because of performance reasons but disabling certain passes would most likely work.

MSVC has a flag we currently ignore /Zo that we could reuse. But maybe it's misleading to disable optimizations with that flag?

thoughts about this approach?

**Comment by pogo59** [comment]
Fussing with the set of passes would be appropriate for an `-Og` optimization level, but not when you add `-g` or similar to the compilation.  The principle is that codegen changes should be controlled by codegen options, and not just because you ask for debug info.

**Comment by tru** [comment]
> Fussing with the set of passes would be appropriate for an `-Og` optimization level, but not when you add `-g` or similar to the compilation. The principle is that codegen changes should be controlled by codegen options, and not just because you ask for debug info.

I agree with this. I am just trying to figure out how best to map this to the clang-cl options. And to me /Zo sounds pretty similar: https://docs.microsoft.com/en-us/cpp/build/reference/zo-enhance-optimized-debugging

there is no /Og for msvc. That said we could create one and call it clang-cl specific - or just tell people to use -Og and disable these passes for the windows target with that flag.

**Comment by dwblaikie** [comment]
> Without changing that LLVM design choice or the CodeView format, the only option we have left is to heroically search the instruction stream for materializations of the constant that overlap with the range over which the variable is known to be that constant.

How's this compare to the DWARF emission, then?

**Comment by rnk** [comment]
I don't think we should allow `/Zo` to control optimization settings. I think it is intended to be a debug info setting, not an optimization setting. I believe that `/Zo` is meant to directly control whether `S_DEFRANGE*` records are used for variable locations or the `S_BPREL` / `S_REGREL` records are used. You can think of it as toggling between LLVM dbg.declare and LLVM dbg.value debug info.

Right now Clang maps `-Og` to `-O1` unless I'm misremembering, or people at least use -O1 as a crutch for lack of -Og. SROA is a pretty core optimization. Nothing fires without it, so it has to be in O1. SCCP I'm not sure of, and could be removed if we like.

It occurs to me that clang-cl maps `/O1` to `-Os`, so there's no -O1 equivalent. That's something worth thinking about. I believe -O3 is also not exposed.

**Comment by tru** [comment]
So it seems to me that we have few ways forward:

* Disable SCCP under `O1` - maybe only when target is windows-msvc, since this pass doesn't negatively effect DWARF, this requires that we re-map `/O1` to `-O1` instead of `-Os`, not sure what the implications for current users will be then.
* Disable SCCP under `Og` and expose `-Og` as `/Og`.
* Add a flag to disable SCCP specifically - maybe `-mllvm -enable-sccppass=false` or something like that?

For my part - I rather add the disabling of this pass into something like `-Og` since there might be other passes in the future that is better disabled for debugable binaries, and it would be easier to continue to improve `Og` instead of manually maintaining the passes to disable. That said - I can work with both and even maintain a local patch to remove this. But I think this might have greater value to others as well since it impacts anyone using windbg and clang-cl.

Additionally it sounds like we want to harmonize the flags between `clang` and `clang-cl` so that `/Ox` maps directly to `-Ox`. But I guess this comes close to a discussion about what `clang-cl` really should do with it flags? Is it trying to map what `cl` is doing the best or is it better that it maps to what a clang developer / user expects?

Another question that was asked earlier that I didn't really understand - does anyone know why `this` is removed from the lambda scope?

**Comment by zmodem** [comment]
> How's this compare to the DWARF emission, then?

I'm curious about this too.

**Comment by tru** [comment]
> Another possible way forward: for a variable that only gets set once, we could maybe just emit it into the debug info as a constant? It is possible to encode the value of a constant in debug info. But not sure a debugger would actually show that the way we want.

Happy to test this - but I am not that well versed in the relevant code - if you don't mind giving me some pointers I can give it a try next week.


> The lamba doesn't actually capture anything, so "this" doesn't contain any data. So the 1-byte empty alloca was optimized out.

aha! Thanks for the insight!

**Comment by efriedma-quic** [comment]
> > Another possible way forward: for a variable that only gets set once, we could maybe just emit it into the debug info as a constant? It is possible to encode the value of a constant in debug info. But not sure a debugger would actually show that the way we want.
> 
> Happy to test this - but I am not that well versed in the relevant code - if you don't mind giving me some pointers I can give it a try next week.

If you were going to try to implement this, the code to emit constants is CodeViewDebug::emitConstantSymbolRecord.  Probably CodeViewDebug::collectVariableInfoFromMFTable would need to pass it the relevant information, somehow?

If you just want to test how a debugger reacts... maybe see what MSVC generates for `int f() { static const int abcd = 0x123456; return abcd; }` with optimizations enabled?  (Apparently, there's a target-independent LLVM bug that makes it throw away the value of the constant in this case, but maybe MSVC doesn't.)

**Comment by rnk** [comment]
I'm optimistic that variables set to a constant once can be represented with S_CONSTANT records. I've seen MSVC emit S_CONSTANT records in function & block scope, so I think it can be done.

**Comment by dwblaikie** [comment]
The DWARF backend optimizes the description of variables that take a constant value for their entire lifetime (it can use a singular location description, rather than a fragmented one)- I'm still not understanding why CV's backend has distinct challenges in this regard (now that I understand that CV can't handle varying locations like DWARF can - which is noteworthy, but there's still a lot of overlap with DWARF) - basically CV should be doing the same things DWARF is, but throwing out any location that doesn't cover the entire scope or is in multiple locations across that scope (ie: any time DWARF would use a location list, rather than a single location description)?

And in the above example, if the increment is optimized away, then there's only one value (1235) that the variable has for every instruction in its scope?

**Comment by rnk** [comment]
CodeView has the equivalent of location lists. It can handle varying locations across different PCs. To the best of our knowledge, there is no way to represent a constant variable value such as `123`. The variable value must be in a machine location, a register or memory location. Fixing this isn't impossible, it just requires a debugger update from Microsoft.

**Comment by tru** [comment]
Hmm - I am confused what you are saying in your latest comments here @rnk - above you say that we could perhaps represent constants with `S_CONSTANT` records - and in your comment above you say that it is not possible to do represent values as constants. Do you mean that `S_CONSTANT` needs the WinDbg changes before we can test that or are these comments talking about different things?

Anyone know what the process of suggesting a change in Visual Studio is?

**Comment by rnk** [comment]
Sorry, I see the confusion. What I mean is that there is no way to say that a variable is a constant over some range of PCs. There is no way for a constant to be part of a location list. So, we have no way to track the initial value of y in this code:
```
int y = 123; // mem2reg will delete the store and produce something like dbg.value(i32 123, !y)
doStuff();
for (; y < n; ++y) // phi inserted at entry between 123 and backend value
  doMoreStuff();
```

If the debugger stops at doStuff(), in DWARF, we could say that y is 123. In CodeView, we cannot yet.

As a special case, if y was never modified, we could use the `S_CONSTANT` record to pretend that the user wrote `static const int y = 123;`. The value would be available over its entire scope. So we can make this work:
```
int y = 123;
doStuff();
for (int x = 0; x < n; ++x)
  doMoreStuff();
```
That's what I thought @efriedma-quic suggested in https://github.com/llvm/llvm-project/issues/55958#issuecomment-1152680452.

**Comment by dwblaikie** [comment]
> What I mean is that there is no way to say that a variable is a constant over some range of PCs. There is no way for a constant to be part of a location list. 

Ah, OK. 

What does MSVC do with the simple cases? (`{ int x = 42; func(); }` (what's the location of `x` during the call to `func()`)) does it do the "pretend the variable was declared const" or does it drop the location entirely, or something else?)

The original issue here had a constant value - and DWARF detects this and uses a single-location description with that constant value. It doesn't seem too weird for CV to do the same, modulo slight abuse of the CV format (describing the variable as a constant when it isn't in the original source), maybe, to provide such a description.

**Comment by rnk** [comment]
In your case, with optimizations, MSVC will not be able to report the value of `x`. However, if `x` was used, they probably could: `int x = 42; func(x);` LLVM, on the other hand, would not be able to, because we immediately constant propagate 42 into the call to func, and there is no connection between the original variable and that argument.

**Comment by tru** [comment]
Hi!

I am once again poking at this one, but I think I need some pointers. As a first test I am just trying to emit `S_CONSTANT` records for variables that has empty `DefRanges`, I know this isn't fully correct, but I wanted to start there just to see how MSVC would behave.

The IR looks like this:

```
define dso_local noundef i32 @main() local_unnamed_addr #0 !dbg !19 {
  call void @llvm.dbg.value(metadata i32 5, metadata !23, metadata !DIExpression()), !dbg !26
  call void @llvm.dbg.value(metadata i32 6, metadata !24, metadata !DIExpression()), !dbg !26
  ret i32 0, !dbg !27
}
...

!23 = !DILocalVariable(name: "i", scope: !19, file: !1, line: 9, type: !12)
!24 = !DILocalVariable(name: "y", scope: !19, file: !1, line: 10, type: !12)
```

Those get represented as a `DIVar` in `emitLocalVariableList`. My first question is how I access the value for the variables in the `DIVar`? (`metadata i32 5` for example). I can't find it in there.

Meanwhile, I just emitted S_CONSTANT records with dummy values which lead me to my second question. These values seem to end up in the global variable stream even if I emit them in the same place as the S_LOCAL does. I have tried to understand how the mapping between an `MCSymbol` and the stream it ends up in works - but can't find that information encoded anywhere. Any pointers are appreciated.

I tried hard to get clang to emit S_CONSTANT's in the code to have something to compare to but was not successful.

**Comment by jmorse** [comment]
Heya @tru,

> Those get represented as a DIVar in emitLocalVariableList. My first question is how I access the value for the variables in the DIVar? (metadata i32 5 for example). I can't find it in there.

By design (for better or worse) there's no way to map from a DILocalVariable to the intrinsics that use it in the IR, alas -- you have to search through the instructions to find variable assignments first. During ASM printing, the DbgEntityHistoryCalculator utility class does that and produces a series of ranges with the variable location. DWARF produces location lists with that information, I don't know about CodeView though.

(I've assumed you're interested in finding variable values / locations during assembly printing rather than in generic LLVM-IR, possibly I've guessed wrong though).

**Comment by tru** [comment]
> Heya @tru,
> 
> By design (for better or worse) there's no way to map from a DILocalVariable to the intrinsics that use it in the IR, alas -- you have to search through the instructions to find variable assignments first. During ASM printing, the DbgEntityHistoryCalculator utility class does that and produces a series of ranges with the variable location. DWARF produces location lists with that information, I don't know about CodeView though.
> 
> (I've assumed you're interested in finding variable values / locations during assembly printing rather than in generic LLVM-IR, possibly I've guessed wrong though).

Yeah so the high-level of what I want to do is to emit S_CONSTANT entries into the ASM stream for variables that have been removed by mem2reg. For example the following code:

```
int a = 5;
foo(a);
```

would remove the variables a and b and just leave the following `llvm.dbg.value`:

```
call void @llvm.dbg.value(metadata i32 5, metadata !23, metadata !DIExpression()), !dbg !26
```

So I would have to find the `5` somehow in order to emit the right instruction into the PDB.

I looked at `llvm::calculateDbgEntityHistory()` but can't say that I am too familiar with the code yet to understand how set up the correct input variables there to find the right one.

I'll continue to poke at this - but some pointers would definitely be helpful.

**Comment by jmorse** [comment]
A brief overview then -- the dbg.values are largely left unchanged through most of the passes, but become DBG_VALUE instructions in instruction selection, recording the same information. The LiveDebugValues pass propagates variable location information to every block. DbgEntityHistoryCalculator produces a collection of instruction-to-instruction ranges for variable locations. What's probably unclear though is that this is done in the base class of CodeViewDebug, `DebugHandlerBase`, where the variable location ranges are stuffed into a `DbgValues` container.

Looking at CodeViewDebug (which I'm not generally familiar with), it looks like `collectVariableInfo` moves location information from the base-class container into containers with CodeView specific field names. It looks like `CodeViewDebug::calculateRanges` transforms the instruction-to-instruction ranges and corresponding DBG_VALUE instructions, into CodeView specific information. That's the last place you can access the constant value of the variable (the first operand of the DBG_VALUE will be a constant "5" in your example), and so presumably where that would need to become some CodeView data representation.

Hope that helps!

**Comment by tru** [comment]
Thanks @jmorse! This was very helpful and I was able to figure it out. With this test program:

```
int t(int i)
{
    return i * 64;
}

int main(int argc, char** argv)
{
    int i = 123;
    int b = t(i);

    return b;
}
```

Building with clang-cl with -O2 this is what happens before my patch:

![image](https://user-images.githubusercontent.com/35614/203286608-aedf5153-d980-48ff-9a06-0c0afcd4d03d.png)

And this is after:

![image](https://user-images.githubusercontent.com/35614/203286656-4300fc0c-72ae-4f68-a3f8-c57aa1f74ae4.png)

Now, these variables show up as global variables because I haven't figured out how to add `S_CONSTANT` record as a local record instead of a global one. Maybe you could point me in the right direction here @rnk?

This code is probably not making the right decisions on when to emit them as constants yet, but here is a draft of the changes I made: https://github.com/tru/llvm-project/commit/21e1c97af52875e13c94a0dfc0ee228e66a14a8b

Happy to get some feedback on if I am moving in the right direction or not.

**Comment by tru** [comment]
Ah never mind - figured it out - the stream selection is happening in LLD of course. Now I just need to figure out how to select if a symbol is global or local :)

**Comment by efriedma-quic** [comment]
Not completely sure what you mean by "global or local".  If the symbol is in the wrong scope, that's a bug which should be fixable.  If you mean the storage duration affects the display somehow, I'm not sure that's something we can control; the patch is basically pretending the variable is a "static" variable to work around the limitations in the encoding.

**Comment by efriedma-quic** [comment]
Looking at your draft more closely, you probably need to use CodeViewDebug::getFullyQualifiedName() to get the scope right.

**Comment by tru** [comment]
> Looking at your draft more closely, you probably need to use CodeViewDebug::getFullyQualifiedName() to get the scope right.

I tried this and it didn't work - that ended up inserting a record with `main::i` and this doesn't show up in Visual Studio at all.

We have now done quite a bit of testing internally and this patch definitely improves debugging for us when enabling optimizations.

I have submitted a patch for review here: https://reviews.llvm.org/D138995
